app/src/main/java/org/mozilla/tv/firefox/webrender/YouTubeBackHandler.kt

```diff
@@ -59,7 +61,34 @@ class YouTubeBackHandler(
         engineView?.evalJS(javascript = shouldWeExitPage, callback = backOrMoveFocus)
     }
 
-    private fun goBackBeforeYouTube() {
-        engineView?.handleYoutubeBack()
+    /**
+     * We only want to add indexes to the list that are a forward navigation from non-YouTube site
+     * to YouTube. If you hit back from a non-YouTube site to YouTube, we do not want to save the index
+     * of that non-YouTube site. The web history list is only updated after loading has completed,
+     * which is why this function is called from onLoadingStateChanged callback.
+     */
+    fun updateUrlIndexList() {
+        if (goingToYouTube) {
+            backForwardUrlList = engineView!!.backForwardList
+            if (indexToAdd < backForwardUrlList.currentIndex)
+                urlIndexList.add(indexToAdd)
+        }
+    }
+
+    fun saveIndex(url: String) {
+        // If you are visiting a YouTube url from a non-YouTube url, save the index
+        if (!goingToYouTube && url.isUriYouTubeTV) {
+            goingToYouTube = true
+            backForwardUrlList = engineView!!.backForwardList
+            indexToAdd = backForwardUrlList.currentIndex
+        } else if (!url.isUriYouTubeTV) goingToYouTube = false
     }
+
+    private fun getIndexToGoBackTo(): Int {
+        return if (urlIndexList.isNotEmpty())
+            urlIndexList.removeAt(urlIndexList.lastIndex)
+        else 0
+    }
+
+    private fun goBackBeforeYouTube() = engineView!!.handleYoutubeBack(getIndexToGoBackTo())
```

Comment:
This logic has gotten a little hard to follow IMO.  I'm seeing if I can work out any other way we can format this that might be easier to read, but it's a hard problem.

Reply:
This is the same logic as your code, but I took some time to try to improve its readability.  A lot of this is subjective, so feel free to take some / all / or none of this advice.

```kotlin
private val preYouTubeIndexHistory: MutableList<Int> = mutableListOf()
private var currentPreYouTubeIndex: Int? = null

fun onUrlChanged(url: String) {
    if (!url.isUriYouTubeTV) currentPreYouTubeIndex = null
    else if (currentPreYouTubeIndex == null) {
        // Store the current (pre-YouTube) backForwardIndex when the URL first changes to YouTube
        currentPreYouTubeIndex = engineView!!.backForwardList.currentIndex
    }
}

/**
 * We only want to add indexes to the list that are a forward navigation from non-YouTube site
 * to YouTube. If you hit back from a non-YouTube site to YouTube, we do not want to save the index
 * of that non-YouTube site. The web history list is only updated after loading has completed,
 * which is why this function is called from onLoadingStateChanged callback.
 */
fun onLoadComplete() {
    // Don't store anything if we aren't navigating to YouTube
    val preYouTubeIndex = currentPreYouTubeIndex ?: return

    val navigationWasForward = preYouTubeIndex < engineView!!.backForwardList.currentIndex
    if (navigationWasForward) {
        preYouTubeIndexHistory += preYouTubeIndex
    }
}

private fun getIndexToGoBackTo(): Int {
    return if (preYouTubeIndexHistory.isEmpty()) 0
    else preYouTubeIndexHistory.removeAt(preYouTubeIndexHistory.lastIndex)
}

private fun goBackBeforeYouTube() = engineView!!.handleYoutubeBack(getIndexToGoBackTo())
```

Changes made:
- Made `backForwardUrlList` a local variable (as we were updating it before reading anyway)
- Renamed `urlIndexList` and `indexToAdd` to `preYouTubeIndexHistory` and `currentPreYouTubeIndex` to try and communicate more about what they represented
- Merged `indexToAdd` and `goingToYouTube` concepts into a nullable Int
- Renamed `updateUrlIndexList` and `saveIndex` to show that they are callbacks. This keeps us from having to read the other file to understand when they are called
- Moved `saveIndex` above `updateUrlIndexList` (it's called first, so the file can now be read more top to bottom)
- Restructured some conditional statements to complete their simpler checks first, to make them easier to follow

Again, a lot of this is subjective.  This would be my advice, but I'm wrong a lot!